C7Engine/EntryPoints/UnitInteractions.cs

```diff
@@ -84,85 +75,21 @@ private static MapUnit UnitWithAvailableActions(MapUnit unit)
-        // powerful moveUnit function that can accept non-neighboring tiles and/or do things like rebase air units. Also direction should be an
-        // enum or something.
         public static void moveUnit(string guid, TileDirection dir)
         {
-            GameData gameData = EngineStorage.gameData;
-            //This is inefficient, perhaps we'll have a map someday.  But with three units,
-            //we'll survive for now.
-            foreach (MapUnit unit in gameData.mapUnits)
-            {
-                if (unit.guid == guid)
-                {
-                    (int dx, int dy) = dir.toCoordDiff();
-                    Tile newLoc = gameData.map.tileAt(dx + unit.location.xCoordinate, dy + unit.location.yCoordinate);
-                    if (newLoc != null && newLoc.IsLand()) {
-                        if (unit.movementPointsRemaining > 0) {
-                            if (!unit.location.unitsOnTile.Remove(unit))
-                                throw new System.Exception("Failed to remove unit from tile it's supposed to be on");
-                            newLoc.unitsOnTile.Add(unit);
-                            unit.location = newLoc;
-                            unit.facingDirection = dir;
-                            unit.movementPointsRemaining -= newLoc.overlayTerrainType.movementCost;
-                            unit.isFortified = false;
-                            EngineStorage.animTracker.startAnimation(unit, MapUnit.AnimatedAction.RUN, null);
-                        }
-                    }
-
-                    break;
-                }
-            }
+            new MsgMoveUnit(guid, dir).send();
         }
 
-        /**
-         * I'd like to enhance this so it's like Civ4, where the hold action takes
-         * the unit out of the rotation, but you can change your mind if need be.
-         * But for now it'll be like Civ3, where you're out of luck if you realize
-         * that unit was needed for something; that also simplifies things here.
-         **/
         public static void holdUnit(string guid)
         {
-            GameData gameData = EngineStorage.gameData;
-            //This is inefficient, perhaps we'll have a map someday.  But with three units,
-            //we'll survive for now.
-            foreach (MapUnit unit in gameData.mapUnits)
-            {
-                if (unit.guid == guid)
-                {
-                    Console.WriteLine("Found matching unit with guid " + guid + " of type " + unit.GetType().Name + "; settings its movement to zero");
-                    unit.movementPointsRemaining = 0;
-                    return;
-                }
-            }
-            Console.WriteLine("Failed to find a matching unit with guid " + guid);
+            new MsgSkipUnitTurn(guid).send();
         }
 
-        public static void waitUnit(string guid)
+        public static void waitUnit(GameData gameData, string guid)
```

Comment:
Shouldn't this one also create a message to the engine?  I've been using the wait action right away to make the camera center on my units (which your other PR will make unnecessary), and noticed it isn't working anymore, and that later on if I select a fortified unit, wait it, and then fortify it, the turn doesn't register as being over (whereas if I just select it and fortify it, it does).  Haven't dug all the way into why (started with functional testing), but I suspect it may be related to there not being a MessageToEngine for wait whereas there are for other actions?

Reply:
I'd rather not require an engine message to wait units because in multiplayer we either have to (1) sync that message with the server and suffer the latency or (2) carve out an exception where that message, unlike all the others, is not synchronized. I think it's better to think of the methods that wait units as a little utility library that the engine provides for the UI independent of the gameplay stuff.

As for why the UI doesn't always recognize when all units have been moved, we'll need to go over all that logic carefully at some point and work it out. I've noticed the same thing happening occasionally even though I don't use the wait command. There's a lot of room for edge cases in there since unit fortification happens asynchronously and we are (or at least I am) trying to allow for units to be fortified and selected at the same time. For example (I don't know if you've seen it) I had to add that awkward KeepCSUWhenFortified variable to the Game class. It's necessary so the autoselector doesn't prevent the player from selecting fortified units but adds another piece of mutable state that might be interacting badly with other things.

Reply:
I think there's a challenge in that the wait action requires some state to be stored to know which actions have been waited.  That does not need to be sychronized in multiplayer, but waited units aren't selected until all other ones have been selected, so the fact that a unit has been waited needs to be stored somewhere (and I added that when I first added the functionality).

I saw the KeepCSUWhenFortified, but don't fully understand the goal.  Is the idea like in Civ4 where you can select a fortified unit and it isn't automatically un-fortified?  My understanding is that's important in Civ4 because fortification bonuses grow over several turns, but it isn't the case in Civ3.  That is something we could add, although I hadn't seen it on the radar anywhere.

I guess the bigger view, though, is that this is a regression; Wait works properly in Aztec, but with this change it won't work properly in Babylon.

Reply:
The wait command was broken because it wasn't autoselecting the next unit after the currently selection was set to wait. I must have carelessly deleted the call to GetNextAutoselectedUnit, I reinserted it and now waiting works again.

KeepCSUWhenFortified is to deal with the fact that fortification happens asynchronously and I also want to allow for units to be selected while fortified. The code originally called GetNextAutoselectedUnit after the fortify command was issued, but that doesn't work now that fortification is async b/c the unit won't actually be fortified until the message is processed. So I deleted that call to GetNextAutoselectedUnit and inserted some code that's run every tick that checks if the currently selected unit is fortified and, if so, calls it then. But that prevents fortified units from being selected so I had to add KeepCSUWhenFortified to stop that from happening as needed. I don't really like KeepCSUWhenFortified and we can probably ditch it later when/if we have a way for the engine to let the UI know the fates of messages, i.e. when they've been processed or if they've been rejected.

Reply:
Wait looks good now.

It does feel a bit like the block in line 169 of Game.cs is becoming a bit complicated.  After unwinding the parentheses, I'm pretty sure it means:

Get the next unit if:

 - There is a unit selected, AND
 - Either:
    - It's fortified and we should get the next unit when it's fortified (Civ3 default)
    - Both:
      - It's out of movement points OR it's dead (in other words, it's out of actions)
      - It's not animating

We might want to make some helper methods to reduce the complexity of that conditional (especially if my understanding is not accurate).

Is the concern with "that doesn't work now that fortification is async b/c the unit won't actually be fortified until the message is processed" that it might return the already-selected unit because it hasn't processed the command?  It feels like an early sign of network syncing nightmares to come.  Although I wonder if it's possible to bypass the concern somehow?  E.g. if it were going over a real network, and using TCP, I'm 99% sure that TCP enforces the order of messages.